sql: Internal Executor does not work properly with Virtual Index

**Problem** 
When including the INDEX field when defining virtual schema tables in crdb_internal.go and calling it through a built-in, it causes a panic

**To Reproduce**

1. Apply this diff
```
diff --git a/pkg/sql/crdb_internal.go b/pkg/sql/crdb_internal.go
index 28502a8986..0ab8504bd0 100644
--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -2243,7 +2243,8 @@ CREATE TABLE crdb_internal.create_type_statements (
   descriptor_id      INT,
   descriptor_name    STRING,
   create_statement   STRING,
-  enum_members       STRING[] -- populated only for ENUM types
+  enum_members       STRING[], -- populated only for ENUM types
+INDEX (descriptor_id)
 )
 `,
     populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) error {
```
2. start a server
2. run `create type greeting as enum ('hi');`
3. run `SHOW CREATE ALL TYPES;`

**Expected behavior**
This should not panic since the enums are being properly populated

**Additional data / screenshots**
* ERROR: [n1,client=127.0.0.1:64058,hostssl,user=demo,intExec=crdb_internal.show_create_all_types] a panic has occurred!
* runtime error: index out of range [0] with length 0
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | runtime.goPanicIndex
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:90
*   | github.com/cockroachdb/cockroach/pkg/sql.virtualSchemaTable.getIndex
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:220
*   | github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:628
*   | github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491
*   | runtime.goexit
*   |   /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: index out of range [0] with length 0
* Error types: (1) *withstack.withStack (2) runtime.boundsError
*
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 7396 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0009ec090, {0x99ef238, 0xc001288740})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:254 +0xaa
panic({0x80e4e40, 0xc001568360})
        /usr/local/opt/go/libexec/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/sql.virtualSchemaTable.getIndex(...)
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:220
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1({0x99555e0, 0xc000674400})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:628 +0x7f9
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3({0x99ef238, 0xc001288740})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x130
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:481 +0x445


**Environment:**
 - CockroachDB version [e.g. 2.0.x]
 - Server OS: [e.g. Linux/Distrib]
 - Client app [e.g. `cockroach sql`, JDBC, ...]

**Additional context**
What was the impact?

Add any other context about the problem here.

--- a/pkg/sql/BUILD.bazel
+++ b/pkg/sql/BUILD.bazel
@@ -584,6 +584,7 @@ go_test(
         "//pkg/sql/catalog/descs",
         "//pkg/sql/catalog/lease",
         "//pkg/sql/catalog/multiregion",
+        "//pkg/sql/catalog/schemadesc",
         "//pkg/sql/catalog/systemschema",
         "//pkg/sql/catalog/tabledesc",
         "//pkg/sql/catalog/typedesc",

--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -2233,6 +2233,54 @@ CREATE TABLE crdb_internal.builtin_functions (
 	},
 }
 
+func writeCreateTypeDescRow(
+	db catalog.DatabaseDescriptor,
+	sc string,
+	typeDesc catalog.TypeDescriptor,
+	addRow func(...tree.Datum) error,
+) (written bool, err error) {
+	switch typeDesc.GetKind() {
+	case descpb.TypeDescriptor_ENUM:
+		var enumLabels tree.EnumValueList
+		enumLabelsDatum := tree.NewDArray(types.String)
+		for i := 0; i < typeDesc.NumEnumMembers(); i++ {
+			rep := typeDesc.GetMemberLogicalRepresentation(i)
+			enumLabels = append(enumLabels, tree.EnumValue(rep))
+			if err := enumLabelsDatum.Append(tree.NewDString(rep)); err != nil {
+				return false, err
+			}
+		}
+		name, err := tree.NewUnresolvedObjectName(2, [3]string{typeDesc.GetName(), sc}, 0)
+		if err != nil {
+			return false, err
+		}
+		node := &tree.CreateType{
+			Variety:    tree.Enum,
+			TypeName:   name,
+			EnumLabels: enumLabels,
+		}
+		return true, addRow(
+			tree.NewDInt(tree.DInt(db.GetID())),       // database_id
+			tree.NewDString(db.GetName()),             // database_name
+			tree.NewDString(sc),                       // schema_name
+			tree.NewDInt(tree.DInt(typeDesc.GetID())), // descriptor_id
+			tree.NewDString(typeDesc.GetName()),       // descriptor_name
+			tree.NewDString(tree.AsString(node)),      // create_statement
+			enumLabelsDatum,
+		)
+	case descpb.TypeDescriptor_MULTIREGION_ENUM:
+		// Multi-region enums are created implicitly, so we don't have create
+		// statements for them.
+		return false, nil
+	case descpb.TypeDescriptor_ALIAS:
+		// Alias types are created implicitly, so we don't have create
+		// statements for them.
+		return false, nil
+	default:
+		return false, errors.AssertionFailedf("unknown type descriptor kind %s", typeDesc.GetKind().String())
+	}
+}
+
 var crdbInternalCreateTypeStmtsTable = virtualSchemaTable{
 	comment: "CREATE statements for all user defined types accessible by the current user in current database (KV scan)",
 	schema: `
@@ -2243,53 +2291,37 @@ CREATE TABLE crdb_internal.create_type_statements (
   descriptor_id      INT,
   descriptor_name    STRING,
   create_statement   STRING,
-  enum_members       STRING[] -- populated only for ENUM types
+  enum_members       STRING[], -- populated only for ENUM types
+  INDEX (descriptor_id)
 )
 `,
 	populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) error {
 		return forEachTypeDesc(ctx, p, db, func(db catalog.DatabaseDescriptor, sc string, typeDesc catalog.TypeDescriptor) error {
-			switch typeDesc.GetKind() {
-			case descpb.TypeDescriptor_ENUM:
-				var enumLabels tree.EnumValueList
-				enumLabelsDatum := tree.NewDArray(types.String)
-				for i := 0; i < typeDesc.NumEnumMembers(); i++ {
-					rep := typeDesc.GetMemberLogicalRepresentation(i)
-					enumLabels = append(enumLabels, tree.EnumValue(rep))
-					if err := enumLabelsDatum.Append(tree.NewDString(rep)); err != nil {
-						return err
-					}
-				}
-				name, err := tree.NewUnresolvedObjectName(2, [3]string{typeDesc.GetName(), sc}, 0)
-				if err != nil {
-					return err
-				}
-				node := &tree.CreateType{
-					Variety:    tree.Enum,
-					TypeName:   name,
-					EnumLabels: enumLabels,
-				}
-				if err := addRow(
-					tree.NewDInt(tree.DInt(db.GetID())),       // database_id
-					tree.NewDString(db.GetName()),             // database_name
-					tree.NewDString(sc),                       // schema_name
-					tree.NewDInt(tree.DInt(typeDesc.GetID())), // descriptor_id
-					tree.NewDString(typeDesc.GetName()),       // descriptor_name
-					tree.NewDString(tree.AsString(node)),      // create_statement
-					enumLabelsDatum,
-				); err != nil {
-					return err
-				}
-			case descpb.TypeDescriptor_MULTIREGION_ENUM:
-				// Multi-region enums are created implicitly, so we don't have create
-				// statements for them.
-			case descpb.TypeDescriptor_ALIAS:
-			// Alias types are created implicitly, so we don't have create
-			// statements for them.
-			default:
-				return errors.AssertionFailedf("unknown type descriptor kind %s", typeDesc.GetKind().String())
-			}
-			return nil
+			_, err := writeCreateTypeDescRow(db, sc, typeDesc, addRow)
+			return err
 		})
+	},
+	indexes: []virtualIndex{
+		{
+			populate: func(
+				ctx context.Context,
+				constraint tree.Datum,
+				p *planner,
+				db catalog.DatabaseDescriptor,
+				addRow func(...tree.Datum) error,
+			) (matched bool, err error) {
+				d := tree.UnwrapDatum(p.EvalContext(), constraint)
+				if d == tree.DNull {
+					return false, nil
+				}
+				id := descpb.ID(tree.MustBeDInt(d))
+				scName, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id)
+				if err != nil || typDesc == nil {
+					return false, err
+				}
+				return writeCreateTypeDescRow(db, scName, typDesc, addRow)
+			},
+		},
 	},
 }
 

--- a/pkg/sql/logictest/testdata/logic_test/crdb_internal
+++ b/pkg/sql/logictest/testdata/logic_test/crdb_internal
@@ -762,6 +762,15 @@ SELECT * FROM crdb_internal.create_type_statements
 
 # Test the virtual index as well.
 
+query ITTITTT
+SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = (('enum1'::regtype::oid::int) - 100000)::oid
+----
+54  test  public  66  enum1  CREATE TYPE public.enum1 AS ENUM ('hello', 'hi')  {hello,hi}
+
+query ITTITTT
+SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = 'foo'::regclass::oid
+----
+
 statement ok
 SET application_name = "test_txn_statistics"
 

--- a/pkg/sql/logictest/testdata/logic_test/create_statements
+++ b/pkg/sql/logictest/testdata/logic_test/create_statements
@@ -289,7 +289,8 @@ CREATE TABLE crdb_internal.create_type_statements (
    descriptor_id INT8 NULL,
    descriptor_name STRING NULL,
    create_statement STRING NULL,
-   enum_members STRING[] NULL
+   enum_members STRING[] NULL,
+   INDEX create_type_statements_descriptor_id_idx (descriptor_id ASC) STORING (database_id, database_name, schema_name, descriptor_name, create_statement, enum_members)
 )  CREATE TABLE crdb_internal.create_type_statements (
    database_id INT8 NULL,
    database_name STRING NULL,
@@ -297,7 +298,8 @@ CREATE TABLE crdb_internal.create_type_statements (
    descriptor_id INT8 NULL,
    descriptor_name STRING NULL,
    create_statement STRING NULL,
-   enum_members STRING[] NULL
+   enum_members STRING[] NULL,
+   INDEX create_type_statements_descriptor_id_idx (descriptor_id ASC) STORING (database_id, database_name, schema_name, descriptor_name, create_statement, enum_members)
 )  {}  {}
 CREATE TABLE crdb_internal.cross_db_references (
    object_database STRING NOT NULL,

--- a/pkg/sql/pg_catalog.go
+++ b/pkg/sql/pg_catalog.go
@@ -2825,6 +2825,30 @@ func addPGTypeRow(
 	)
 }
 
+func getSchemaAndTypeByTypeID(
+	ctx context.Context, p *planner, id descpb.ID,
+) (string, catalog.TypeDescriptor, error) {
+	typDesc, err := p.Descriptors().GetImmutableTypeByID(ctx, p.txn, id, tree.ObjectLookupFlags{})
+	if err != nil {
+		// If the type was not found, it may be a table.
+		if !(errors.Is(err, catalog.ErrDescriptorNotFound) || pgerror.GetPGCode(err) == pgcode.UndefinedObject) {
+			return "", nil, err
+		}
+		return "", nil, nil
+	}
+
+	sc, err := p.Descriptors().GetImmutableSchemaByID(
+		ctx,
+		p.txn,
+		typDesc.GetParentSchemaID(),
+		tree.SchemaLookupFlags{},
+	)
+	if err != nil {
+		return "", nil, err
+	}
+	return sc.GetName(), typDesc, nil
+}
+
 var pgCatalogTypeTable = virtualSchemaTable{
 	comment: `scalar types (incomplete)
 https://www.postgresql.org/docs/9.5/catalog-pg-type.html`,
@@ -2906,17 +2930,10 @@ https://www.postgresql.org/docs/9.5/catalog-pg-type.html`,
 				if err != nil {
 					return false, err
 				}
-				typDesc, err := p.Descriptors().GetImmutableTypeByID(ctx, p.txn, id, tree.ObjectLookupFlags{})
-				if err != nil {
-					// If the type was not found, it may be a table.
-					if !(errors.Is(err, catalog.ErrDescriptorNotFound) || pgerror.GetPGCode(err) == pgcode.UndefinedObject) {
-						return false, err
-					}
-					typDesc = nil
-				}
 
-				if typDesc == nil {
-					return false, nil
+				scName, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id)
+				if err != nil || typDesc == nil {
+					return false, err
 				}
 
 				// It's an entry for the implicit record type created on behalf of each
@@ -2952,16 +2969,7 @@ https://www.postgresql.org/docs/9.5/catalog-pg-type.html`,
 					return true, nil
 				}
 
-				sc, err := p.Descriptors().GetImmutableSchemaByID(
-					ctx,
-					p.txn,
-					typDesc.GetParentSchemaID(),
-					tree.SchemaLookupFlags{},
-				)
-				if err != nil {
-					return false, err
-				}
-				nspOid = h.NamespaceOid(db.GetID(), sc.GetName())
+				nspOid = h.NamespaceOid(db.GetID(), scName)
 				typ, err = typDesc.MakeTypesT(ctx, tree.NewUnqualifiedTypeName(typDesc.GetName()), p)
 				if err != nil {
 					return false, err

--- a/pkg/sql/virtual_schema_test.go
+++ b/pkg/sql/virtual_schema_test.go
@@ -24,14 +24,18 @@ package sql
 
 import (
 	"bufio"
+	"context"
 	"flag"
 	"fmt"
 	"os"
 	"strings"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemadesc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/require"
 )
 
 var rewriteTables = flag.Bool(
@@ -49,7 +53,9 @@ func TestVirtualSchemas(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	for _, schema := range virtualSchemas {
+	ctx := context.Background()
+
+	for schemaID, schema := range virtualSchemas {
 		_, isSchemaFixable := fixableSchemas[schema.name]
 		if len(schema.undefinedTables) == 0 || !isSchemaFixable {
 			continue
@@ -64,7 +70,7 @@ func TestVirtualSchemas(t *testing.T) {
 			rewriteSchema(schema.name, unimplementedTables)
 		} else {
 			t.Run(fmt.Sprintf("VirtualSchemaTest/%s", schema.name), func(t *testing.T) {
-				for _, virtualTable := range schema.tableDefs {
+				for tableID, virtualTable := range schema.tableDefs {
 					tableName, err := getTableNameFromCreateTable(virtualTable.getSchema())
 					if err != nil {
 						t.Fatal(err)
@@ -74,6 +80,27 @@ func TestVirtualSchemas(t *testing.T) {
 							"Table %s.%s is defined and not expected to be part of undefinedTables",
 							schema.name,
 							tableName,
+						)
+					}
+
+					// Sanity check indexes are all defined.
+					sc, ok := schemadesc.GetVirtualSchemaByID(schemaID)
+					require.True(t, ok)
+					d, err := virtualTable.initVirtualTableDesc(
+						ctx,
+						cluster.MakeClusterSettings(),
+						sc,
+						tableID,
+					)
+					require.NoError(t, err)
+					switch virtualTable := virtualTable.(type) {
+					case *virtualSchemaTable:
+						require.Equalf(
+							t,
+							len(d.GetIndexes()),
+							len(virtualTable.indexes),
+							"number of indexes in description must match number of indexes defined for table %s",
+							d.GetName(),
 						)
 					}
 				}

